src/tir/schedule/primitive/compute_at.cc

```diff
@@ -0,0 +1,584 @@
+      : mod_(mod),
+        block_(GetRef<Block>(block_sref->StmtAs<BlockNode>())),
+        loop_(GetRef<For>(loop_sref->StmtAs<ForNode>())) {}
+
+  IRModule mod_;
+  Block block_;
+  For loop_;
+};
+
+/******** Helper Functions/Classes ********/
+
+/*!
+ * \brief Find a point where the block can be inserted under the loop
+ * \tparam require_all_producers_visited Requires all producer blocks to be present under the loop
+ * \tparam require_all_consumers_visited Requires all consumer blocks to be present under the loop
+ * \param self The schedule state
+ * \param subtrees The subtrees under the loop, among which the insertion points are sought
+ * \param producer_srefs The producer blocks
+ * \param consumer_srefs The consumer blocks
+ * \param block2realize A cache that maps a block to its realize
+ * \return The last position the new block can be inserted onto, and the
+ * producer-consumer-relationship is still satisfied.
+ */
+template <bool require_all_producers_visited, bool require_all_consumers_visited>
+int FindInsertionPoint(
+    const ScheduleState& self, const Array<Stmt>& subtrees, const Array<StmtSRef>& producer_srefs,
+    const Array<StmtSRef>& consumer_srefs,
+    std::unordered_map<const BlockNode*, const BlockRealizeNode*>* block2realize) {
+  ProducerConsumerSplit split =
+      ProducerConsumerSplit::Find(self, subtrees, producer_srefs, consumer_srefs, block2realize);
+  // Step 1. Check if all the producers are visited in the subtrees, if required to
+  if (require_all_producers_visited) {
+    int num_producers = producer_srefs.size();
+    if (split.n_producers_visited < num_producers) {
+      throw NotAllRequiredBlocksAreVisitedError<false>(
+          self->mod, num_producers - split.n_producers_visited, producer_srefs);
+    }
+  }
+  // Step 2. Check if all the consumers are visited in the subtrees, if required to
+  if (require_all_consumers_visited) {
+    int num_consumers = consumer_srefs.size();
+    if (split.n_consumers_visited < num_consumers) {
+      throw NotAllRequiredBlocksAreVisitedError<true>(
+          self->mod, num_consumers - split.n_consumers_visited, consumer_srefs);
+    }
+  }
+  // Step 3. Check if there is at least one index of the position can be inserted into
+  // The valid indices are: (last_producer_position, first_consumer_position]
+  ICHECK(split.last_producer_position < split.first_consumer_position);
+  // Step 4. Return the last valid insertion point
+  return split.first_consumer_position;
+}
+
+/*!
+ * \brief A helper to reconstruct the block scope where the given block is moved under the given
+ * loop, and the given block's induced loop nest is regenerated to satisfy the required region.
+ */
+class ScopeReconstructor : private StmtMutator {
+ public:
+  explicit ScopeReconstructor(Block scope_root, Block block, For loop)
+      : scope_root_(scope_root), block_(block), loop_(loop) {}
+
+  using StmtMutator::operator();
```

Comment:
What does this line mean?

Reply:
it means making this operator public, given it's private inheritance